Feat/containerize application

boekhorstb1 opened the pull request:
NOTE: this is currently dependent on https://github.com/SovereignCloudStack/rookify/pull/23

- added Dockerfile for rookify: python3-rados is installed within the ceph container 
- added Makefile to build and run container of rookify
- added Makefile command to setup venv locally 

ISSUE: 
- ceph container works with python3.6, in centos I could not find python bindings for rados with 3.9

Closes: #2

NotTheEvilOne commented on Dockerfile:
```diff
@@ -0,0 +1,53 @@
+CMD ["sleep", "infinity"]
+
+# Set the ENTRYPOINT to activate the venv and then run the 'rookify' command
+ENTRYPOINT ["/bin/bash", "-c", "source ./bin/activate && exec python -m rookify"]
```
The `rookify` console script installed with the Python `wheel` package is automatically linked to the venv if installed in it. In this case you do not need to source venv beforehand. `ENTRYPOINT ["/app/rookify/bin/rookify"]` will do this for you :)

NotTheEvilOne commented on Makefile:
```diff
@@ -0,0 +1,55 @@
+	${MAKE} setup-venv
+	
+setup-venv:
+	python -m venv ./.venv
```
You may want to use `python -m venv --system-site-packages --symlinks` as in the `Dockerfile`. In that case you do not have to symlink files manually.

brueggemann replied:
I think using --system-site-packages is not a good solution, especially in dev environment, as pip freeze will then list all system packages and you can't really define, what version of a site-package is loaded, if all system site-packages are included in venv

boekhorstb1 replied:
ok for makefile I opened a separate mr here: https://github.com/SovereignCloudStack/rookify/pull/29

NotTheEvilOne commented on Makefile:
```diff
@@ -0,0 +1,55 @@
+	pip install -r requirements.txt
+
+.PHONY: build-container
+TAG ?= 0.0.1
```
I would suggest to name the variable like in the `Dockerfile`. `TAG` seems to be a bit to common (and too specific to git) for me. In fact it's not the Docker tag or related to it anyway.

brueggemann commented on Dockerfile:
```diff
@@ -0,0 +1,53 @@
+# Install Librados:
+# Get rados version from requirements.txt and install it using apt,
+# because librados does not exist as pip package
+RUN apt-get update && apt-get install -y python3-pip python3-rados python3-venv && \
```
I think we should build librados and python3-rados from source, so that we can enforce the librados/python-rados version to the required version

boekhorstb1 replied:
will check if we can use a ceph image. Seems that building librados requires building a whole ceph.

boekhorstb1 commented:
ok in the last commit I just added I changed some bigger things (please review):
- using ceph container now (quincy v17.2.7), note: runs on python3.6 but has rados2.0.0
- added some more information to pyproject.toml because build would not run otherwise
- adapted Makefile a little (this [MR](https://github.com/SovereignCloudStack/rookify/pull/29_) will have most important changes)
- remove rados from requirements.txt because the ceph container will contain it

@NotTheEvilOne @brueggemann

boekhorstb1 commented:
ok so set the tag differently in the makefile as requested. 
Also set container to latest ceph image. Small annoyance (but I think nothing critical): python version in the container is still 3.6. But the build is works. After installing python 3.9 librados.so.2.0.0 is found anyhow.

boekhorstb1 commented:
ok tested some more. Interestingly enough,  the build works fine, but running rookify in the container does not work, because it cannot find rados for python 3.9.
Problem is here: as far as I know, there are simply no ceph containers that use a higher python version than 3.6.
Furthermore I could not find an rpm in the centos repos for python3-rados that offers interfaces for python 3.9 or higher.

b1-luettje commented:
Since we're having a little bit of an chicken-egg-problem by either using the official ceph container image or the official python image we've decided to (try to) target `ubuntu:24.04` (and its corresponding official packages) instead.

Pros:
- Canonical "ensures" that `import rados` should always work after installing the official packages
- 24.04 is a LTS release

Cons:
- No (more) explicit control over ceph and python versions but `3.12` is perfectly fine and `librados2` seems very settled and therefore unlikely to change drastically in the next years. If necessary, we can always include upstream ceph repositories to upgrade ceph/librados2.

NotTheEvilOne commented:
I've now updated `Dockerfile` to reflect the changes @b1-luettje described above and merged main to fix a TODO while working on it. It's ready to get merged in my eyes. Please have a look again @brueggemann and @yeoldegrove :)

yeoldegrove reviewed (approved):
Looks like a sane approach from my side.
Not sure if this already makes sense, but now that we have the container image we can already mention it in the README.
Also, if I just build it and start it, in complains about a missing `config.yaml` and there is no hint on how to create one.
